### PR TITLE
Upgrade Google Auth

### LINKED ISSUE #25: gapi.auth2.init deprecated
Google is shutting down gapi.auth2 support and all users using this library will be rejected

### PR DESCRIPTION
**Pull request by dextermb**
Resolves #25

Upgrades from using `gapi.auth2` to `google.accounts.id` (https://developers.google.com/identity/gsi/web). 

Screenshot 2022-07-22 at 16 33 00

This change, however, is **breaking** as the outputs from the events are different.

* `auth-success` now returns an object with user properties:
  * `name`
  * `email`
  * `picture`
* `auth-failure` now returns a basic string
* `init-error` is no longer used

### CODE REVIEW

**Comment by dextermb** [comment]
Once people are happy with the implementation then I will look at updating the documentation

File: package.json [hunk 1]
```diff
@@ -1,26 +1,27 @@
...
     "eslint": "^8.14.0",
     "mkcert": "^1.5.0",
     "svelte": "^3.44.0",
     "svelte2tsx": "^0.5.9",
-    "typescript": "^4.6.3"
+    "typescript": "^4.6.3",
+    "vite": "^3.0.0"
   },
-  "type": "module",
-  "volta": {
```
Comment (antony) [inline package.json:23]: Why is volta gone?

File: src/lib/google-auth/GoogleAuth.svelte [hunk 2]
```diff
@@ -1,4 +1,8 @@
-<button bind:this={signinCta} {disabled} use:action>
+<button
+  {disabled}
+  use:action
+  on:click={handleClick}
```
Comment (antony) [inline src/lib/google-auth/GoogleAuth.svelte:4]: We should call event handler functions after what they do, not what events they handle

File: src/lib/google-auth/GoogleAuth.svelte [hunk 3]
```diff
@@ -73,22 +78,27 @@
...
+
+  function attachHandler ({ credential }) {
+    if (!credential) {
+      dispatch('auth-failure', { error: 'Unable to get credentials' })
+      return
+    }
+
+    const { name, email, picture } = decode(credential)
+
+    dispatch('auth-success', { user: { name, email, picture } })
```
Comment (antony) [inline src/lib/google-auth/GoogleAuth.svelte:102]: How do we get their Google id to write to their profile?

**Review by antony — CHANGES_REQUESTED** [review]
Looks good, a few improvements for readability and a question

File: src/lib/google-auth/GoogleAuth.svelte (hunk 3 above)
Comment (dextermb) [reply src/lib/google-auth/GoogleAuth.svelte:102]: I believe we can use the `sub` (subject) property of the JWT token:
https://developers.google.com/identity/gsi/web/reference/js-reference#credential

**Comment by antony** [comment]
Note that we don't validate the jwt signature, so it might be better to get rid of the new dependency and just bade64 decode the unencrypted part of the jwt

https://stackoverflow.com/questions/38552003/how-to-decode-jwt-token-in-javascript-without-using-a-library

**Comment by dextermb** [comment]
> Note that we don't validate the jwt signature, so it might be better to get rid of the new dependency and just bade64 decode the unencrypted part of the jwt
> 
> https://stackoverflow.com/questions/38552003/how-to-decode-jwt-token-in-javascript-without-using-a-library

Certainly if you're happy with having the suggested browser function in the codebase. I'll update this

File: package.json (hunk 1 above)
Comment (dextermb) [reply package.json:23]: Removed as part of the upgrade to SvelteKit, I can add it back with a newer version of node

File: src/lib/google-auth/GoogleAuth.svelte (hunk 2 above)
Comment (dextermb) [reply src/lib/google-auth/GoogleAuth.svelte:4]: The click here triggers the Google Sign In popup. Would the function name `triggerPopup` work for you?

File: package.json (hunk 1 above)
Comment (dextermb) [reply package.json:23]: Added back with node `16.15.0` ✅

File: src/lib/google-auth/GoogleAuth.svelte (hunk 2 above)
Comment (dextermb) [reply src/lib/google-auth/GoogleAuth.svelte]: Renamed to `triggerPrompt` ✅

File: src/lib/google-auth/GoogleAuth.svelte (hunk 3 above)
Comment (dextermb) [reply src/lib/google-auth/GoogleAuth.svelte]: Added `sub` renamed as `id` on output of `auth-success` ✅

**Comment by dextermb** [comment]
> Certainly if you're happy with having the suggested browser function in the codebase. I'll update this

Removed `jwt-decode` dependency and added the answer-marked implementation from stackoverflow ✅

**Review by antony — APPROVED** [review]

**Comment by antony** [comment]
lets pnpm link this to the app to try it first before we merge and release.

**Comment by dextermb** [comment]
> lets pnpm link this to the app to try it first before we merge and release.

👋🏻 Hey, just getting back to this today. I've done a decent amount of testing and reading through our internal use-cases.

@antony with minor changes to our login screen I've got this working as we would expect. I've exposed the whole token through the `auth-success` callback.

It's worth noting the GSI does not like running on non-https domains, other than localhost, so this will throw some complications for us and potentially other development workflows.

**Comment by dextermb** [comment]
I've also noticed that `button` is now being respected on our login/sign up screens- e.g. not `width: 100%` by default which is correct. Potentially something else has changed in our platform but to match the same styles as previously seen I updated the styles:

```css
  :global(.social-auth-button button) {
    width: 100%;
    font-family: 'Cabin', sans-serif;
    border-radius: 5px;
  }
```

_But I don't think this is anything to do with this pull request as looking at the code I have not changed anything relating to styles._

**Comment by TheOnlyWayUp** [comment]
Thanks for the PR! I installed from this commit and importing failed, but copying over the GoogleAuth.svelte file and using it like any other component seems to work

:)

**Comment by antony** [comment]
lets also consider this https://github.com/beyonk-adventures/svelte-social-auth/pull/14 when we merge

**Pull request closed without merging**